sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,318 @@
+    let mut public_keys = Vec::new();
+    for endpoint in public_key_endpoints.iter() {
+        let public_key_text = get_text_from_imds(&client, endpoint.as_str(), &imds_session_token)?;
+        // Simple check to see if the text is probably an ssh key.
+        if &public_key_text[0..3] == "ssh" {
+            debug!("{}", &public_key_text.trim_end());
+            public_keys.push(public_key_text.trim_end().to_string())
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid key. Skipping...",
+                &public_key_text.trim_end()
+            );
+            continue;
+        }
+    }
+    ensure!(!public_keys.is_empty(), error::NoValidImdsKeys);
+
+    Ok(public_keys)
+}
+
+/// Returns a list of endpoint urls for the public keys in IMDS. Since IMDS returns the list of
+/// available public keys as '0=my-public-key', we need to strip the index from the list and
+/// embed it into the endpoint url.
+fn get_public_key_endpoints_from_imds(public_key_list: &str) -> Result<Vec<Url>> {
+    let mut public_key_endpoints = Vec::new();
+    for available_key in public_key_list.lines() {
+        let f: Vec<&str> = available_key.split('=').collect();
+        // If f[0] isn't a number, then it isn't a valid index.
+        if f[0].parse::<i32>().is_ok() {
+            let public_key_endpoint = &format!("{}{}/openssh-key", IMDS_PUBLIC_KEY_BASE_URL, f[0]);
+            let public_key_endpoint_url =
+                Url::parse(&public_key_endpoint).context(error::HttpUrlParse {
+                    url: public_key_endpoint,
+                })?;
+            debug!("{}", &public_key_endpoint);
+            public_key_endpoints.push(public_key_endpoint_url);
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid index. Skipping...",
+                &f[0]
+            );
+            continue;
+        }
+    }
+    ensure!(
+        !public_key_endpoints.is_empty(),
+        error::NoValidImdsKeyEndpoints
+    );
+    Ok(public_key_endpoints)
+}
+
+/// Print usage message.
+fn usage() -> ! {
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!("Usage: {} user-data", program_name);
+    process::exit(2);
+}
+
+/// Parses args for the setting key name.
+fn parse_args(mut args: env::Args) -> String {
+    let arg = args.nth(1).unwrap_or_else(|| usage());
+    if arg == "--help" || arg == "-h" {
+        usage()
+    }
+    arg
+}
+
+async fn run() -> Result<()> {
+    let setting_name = parse_args(env::args());
+
+    // TerminalMode::Mixed will send errors to stderr and anything less to stdout.
+    TermLogger::init(
+        LevelFilter::Debug,
+        LogConfig::default(),
+        TerminalMode::Mixed,
+    )
+    .context(error::Logger)?;
+
+    info!("shibaken started");
+
+    info!("Fetching user-data metadata");
+    let metadata_user_data = get_metadata(&setting_name, "user-data").await?;
```

Comment:
Let's re-name this variable to just `user_data`.  `metadata_user_data` is sort of confusing.